libafl/src/monitors/prometheus.rs

```diff
@@ -0,0 +1,361 @@
+        f.debug_struct("PrometheusMonitor")
+            .field("start_time", &self.start_time)
+            .field("client_stats", &self.client_stats)
+            .finish()
+    }
+}
+
+impl<F> Monitor for PrometheusMonitor<F>
+where
+    F: FnMut(String),
+{
+    /// the client monitor, mutable
+    fn client_stats_mut(&mut self) -> &mut Vec<ClientStats> {
+        &mut self.client_stats
+    }
+
+    /// the client monitor
+    fn client_stats(&self) -> &[ClientStats] {
+        &self.client_stats
+    }
+
+    /// Time this fuzzing run stated
+    fn start_time(&mut self) -> Duration {
+        self.start_time
+    }
+
+    fn display(&mut self, event_msg: String, sender_id: u32) {
+        // Update the prometheus metrics
+        // Label each metric with the sender / client_id
+        let corpus_size = self.corpus_size();
+        self.corpus_count
+            .get_or_create(&Labels {
+                client: sender_id,
+                stat: "".to_owned(),
+            })
+            .set(corpus_size);
+        let objective_size = self.objective_size();
+        self.objective_count
+            .get_or_create(&Labels {
+                client: sender_id,
+                stat: "".to_owned(),
+            })
+            .set(objective_size);
+        let total_execs = self.total_execs();
+        self.executions
+            .get_or_create(&Labels {
+                client: sender_id,
+                stat: "".to_owned(),
+            })
+            .set(total_execs);
+        let execs_per_sec = self.execs_per_sec();
+        self.exec_rate
+            .get_or_create(&Labels {
+                client: sender_id,
+                stat: "".to_owned(),
+            })
+            .set(execs_per_sec);
+        let run_time = (current_time() - self.start_time).as_secs();
+        self.runtime
+            .get_or_create(&Labels {
+                client: sender_id,
+                stat: "".to_owned(),
+            })
+            .set(run_time); // run time in seconds, which can be converted to a time format by Grafana or similar
+        let total_clients = self.client_stats().len().try_into().unwrap(); // convert usize to u64 (unlikely that # of clients will be > 2^64 -1...)
+        self.clients_count
+            .get_or_create(&Labels {
+                client: sender_id,
+                stat: "".to_owned(),
+            })
+            .set(total_clients);
+
+        // display stats in a SimpleMonitor format
+        let fmt = format!(
+            "[Prometheus] [{} #{}] run time: {}, clients: {}, corpus: {}, objectives: {}, executions: {}, exec/sec: {}",
+            event_msg,
+            sender_id,
+            format_duration_hms(&(current_time() - self.start_time)),
+            self.client_stats().len(),
+            self.corpus_size(),
+            self.objective_size(),
+            self.total_execs(),
+            self.execs_per_sec()
+        );
+        (self.print_fn)(fmt);
```

Comment:
For this we potentially want the option to stack/join multiple monitors, so that the prometheus monitor doesn't have to print direclty (and could also offer a tui), what do you think @andreafioraldi

Reply:
I was thinking of a monitor that wraps two monitors